### PR TITLE
[SPARK-41354][CONNECT] Add `RepartitionByExpression` to proto

### PR DESCRIPTION
**Pull request by dengziming**
### What changes were proposed in this pull request?
~~1. Implement `DataFrame.repartitionByRange` for scala API~~
2. Add a new relation `RepartitionByExpression`.

### Why are the changes needed?
API coverage

### Does this PR introduce _any_ user-facing change?
Yes, added new API function.

### How was this patch tested?
Unit tests.

### CODE REVIEW

File: connector/connect/common/src/main/protobuf/spark/connect/relations.proto [hunk 1]
```diff
@@ -598,3 +599,14 @@ message ToSchema {
...
+
+  message RepartitionByExpression {
+  // (Required) The input relation.
+  Relation input = 1;
+
+  // (Required) The partitioning expressions
+  repeated Expression partition_exprs = 2;
+
+  // (Optional) number of partitions, must be positive.
+  int32 num_partitions = 3;
```
Comment (grundprinzip) [inline connector/connect/common/src/main/protobuf/spark/connect/relations.proto:611]:
```suggestion
  optional int32 num_partitions = 3;
```

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 2]
```diff
@@ -372,6 +374,22 @@ class SparkConnectPlanner(session: SparkSession) {
     }
   }
 
+  private def transformRepartitionByExpression(
+      rel: proto.RepartitionByExpression): LogicalPlan = {
+    val numPartitionsOpt = if (rel.getNumPartitions == 0) {
```
Comment (grundprinzip) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:379]: with the optional change you can actually use `if (rel.hasNumPartitions) {`

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 3]
```diff
@@ -372,6 +374,22 @@ class SparkConnectPlanner(session: SparkSession) {
...
+      None
```
Comment (grundprinzip) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:380]: please use the actual value in the if branch and `None` in the else

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 4]
```diff
@@ -372,6 +374,22 @@ class SparkConnectPlanner(session: SparkSession) {
...
+    } else {
+      Some(rel.getNumPartitions)
+    }
+    val partitionExpressions = rel.getPartitionExprsList.asScala.map { expr =>
+      transformExpression(expr)
+    }.toSeq
```
Comment (grundprinzip) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:386]:
```suggestion
    val partitionExpressions = rel.getPartitionExprsList.asScala.map(transformExpression).toSeq
```

File: connector/connect/server/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectPlannerSuite.scala [hunk 5]
```diff
@@ -705,4 +705,86 @@ class SparkConnectPlannerSuite extends SparkFunSuite with SparkConnectPlanTest {
       }
     }
   }
+
+  test("RepartitionByExpression") {
```
Comment (grundprinzip) [inline connector/connect/server/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectPlannerSuite.scala:709]: please add a test that would throw an exception during the execution

**Review by grundprinzip — COMMENTED** [review]
Are you going to add the Python side in a follow up?

File: connector/connect/common/src/main/protobuf/spark/connect/relations.proto [hunk 6]
```diff
@@ -57,6 +57,7 @@ message Relation {
     Hint hint = 24;
     Unpivot unpivot = 25;
     ToSchema to_schema = 26;
+    RepartitionByExpression repartition_by_expression = 27;
```
Comment (zhengruifeng) [inline connector/connect/common/src/main/protobuf/spark/connect/relations.proto:60]: shoule be `RepartitionByRange`?  it's not an experssion

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/dsl/package.scala [hunk 7]
```diff
@@ -656,6 +668,56 @@ package object dsl {
             Repartition.newBuilder().setInput(logicalPlan).setNumPartitions(num).setShuffle(true))
           .build()
 
+      def repartition(partitionExprs: Expression*): Relation = {
```
Comment (zhengruifeng) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/dsl/package.scala:671]: is this one related to `repartitionByRange` ?

**Review by zhengruifeng — COMMENTED** [review]
are you going to support both

`def repartition(self, numPartitions: int, *cols: "ColumnOrName") -> "DataFrame"`

and 

`def repartitionByRange(self, numPartitions: int, *cols: "ColumnOrName") -> "DataFrame"`

in this PR?

I'm a bit confused.

**Review by dengziming — COMMENTED** [review]
Thank you for your comments @grundprinzip @zhengruifeng , I have fixed these bugs, and I have decided to implement python API in a following PR since we just have some adjustments after [SPARK-41651: Test parity: pyspark.sql.tests.test_dataframe](https://issues.apache.org/jira/browse/SPARK-41651)
I also changed the title to mention `DataFrame.repartition(Expression*)`

**Review by zhengruifeng — APPROVED** [review]
LGTM

**Comment by zhengruifeng** [comment]
would you mind resolving the conflicts and update the PR description and title (it aims to adds the proto, not for scala API)?

**Comment by dengziming** [comment]
> would you mind resolving the conflicts and update the PR description and title (it aims to adds the proto, not for scala API)?

@zhengruifeng Thank you for you comments, Do you mean change title to `RepartitionByExpression`, I have change it.

**Comment by zhengruifeng** [comment]
sorry, would you minding resolving the conflicts? @dengziming

**Comment by zhengruifeng** [comment]
Merged into master, thank you @dengziming 

BTW, would you minding also adding the repartitionBy* APIs in Client ?

**Comment by dengziming** [comment]
> adding the repartitionBy* APIs in Client ?

Do you mean adding them to python client? yes, I'm working on it.

**Comment by zhengruifeng** [comment]
@dengziming thank you so much
